pkg/catalog/routes.go

```diff
@@ -397,3 +418,75 @@ func listTrafficTargetPermutations(name string, srcServiceList []service.MeshSer
 
 	return trafficPolicies
 }
+
+func (mc *MeshCatalog) getWeightedBackendsForService(svc service.MeshService) []service.WeightedService {
+	weightedSvcs := []service.WeightedService{}
+
+	// Retrieve the weighted clusters from traffic split
+	// Assumes no nested splits for now
+	// Assumes service would only be root once (not sure if this is validated)
+	servicesList := mc.meshSpec.ListTrafficSplitServices()
+	for _, activeService := range servicesList {
+		if activeService.RootService == svc.Name {
+			weightedSvcs = append(weightedSvcs, activeService)
+		}
+	}
+
+	if len(weightedSvcs) == 0 {
+		svc := service.WeightedService{
+			Service: svc,
+			Weight:  constants.ClusterWeightAcceptAll,
+		}
+		weightedSvcs = append(weightedSvcs, svc)
+	}
+	return weightedSvcs
+}
+
+func (mc *MeshCatalog) getRouteWeightedServices(src service.MeshService, dest service.MeshService, trafficTargets map[string]trafficpolicy.TrafficTarget) []trafficpolicy.RouteWeightedServices {
+	routeWeightedSvcs := []trafficpolicy.RouteWeightedServices{}
+
+	weightedBackends := mc.getWeightedBackendsForService(dest)
+
+	routeServicesMap := make(map[*trafficpolicy.HTTPRoute][]service.WeightedService)
+
+	for _, backend := range weightedBackends {
+		hash := hashSrcDstService(src, backend.Service)
+		routes := trafficTargets[hash].HTTPRoutes
+
+		for i, r := range routes {
+			key := getRouteKey(r, routeServicesMap)
+			if key == nil {
+				servicesValue := []service.WeightedService{backend}
+				key = &routes[i]
+				routeServicesMap[key] = servicesValue
+			} else {
+				servicesValue := routeServicesMap[key]
+				servicesValue = append(servicesValue, backend)
+				routeServicesMap[key] = servicesValue
+			}
+		}
+	}
+
+	for route, services := range routeServicesMap {
+		rWS := trafficpolicy.RouteWeightedServices{
+			HTTPRoute:        *route,
+			WeightedServices: services,
+		}
+		routeWeightedSvcs = append(routeWeightedSvcs, rWS)
+	}
+
+	return routeWeightedSvcs
+}
+
+func getRouteKey(route trafficpolicy.HTTPRoute, routeMap map[*trafficpolicy.HTTPRoute][]service.WeightedService) *trafficpolicy.HTTPRoute {
+	var matchingKey *trafficpolicy.HTTPRoute
+
+	for key := range routeMap {
```

Comment:
The map serves no purpose here I believe. 

You just do a linear search on the map to return the key, to later use the key as a map lookup index.
I have not looked into it deeply, but I'd probably suggest to find a way to index `HTTPRoute` by some usable key.